pkg/webhook/elasticquota/quota_topology.go

```diff
@@ -0,0 +1,475 @@
+	"github.com/koordinator-sh/koordinator/pkg/scheduler/plugins/elasticquota/core"
+	utilclient "github.com/koordinator-sh/koordinator/pkg/util/client"
+)
+
+type quotaTopology struct {
+	lock sync.Mutex
+	// quotaInfoMap stores all quota information
+	quotaInfoMap map[string]*core.QuotaInfo
+	// quotaHierarchyInfo stores the quota's all children
+	quotaHierarchyInfo map[string]map[string]struct{}
+
+	client client.Client
+}
+
+func NewQuotaTopology(client client.Client) *quotaTopology {
+	topology := &quotaTopology{
+		quotaInfoMap:       make(map[string]*core.QuotaInfo),
+		quotaHierarchyInfo: make(map[string]map[string]struct{}),
+		client:             client,
+	}
+	topology.quotaHierarchyInfo[extension.RootQuotaName] = make(map[string]struct{})
+	return topology
+}
+
+func (qt *quotaTopology) AddQuota(quota *v1alpha1.ElasticQuota) error {
+	if quota == nil {
+		return fmt.Errorf("AddQuota param is nil")
+	}
+
+	qt.lock.Lock()
+	defer qt.lock.Unlock()
+
+	if _, exist := qt.quotaInfoMap[quota.Name]; exist {
+		return fmt.Errorf("AddQuota quota already exist:%v", quota.Name)
+	}
+
+	if err := qt.fillQuotaDefaultInformation(quota); err != nil {
+		return err
+	}
+
+	if err := qt.validateQuotaSelfItem(quota); err != nil {
+		return err
+	}
+
+	quotaInfo := core.NewQuotaInfoFromQuota(quota)
+	if err := qt.validateQuotaTopology(quotaInfo); err != nil {
+		return err
+	}
+
+	qt.quotaInfoMap[quotaInfo.Name] = quotaInfo
+	qt.quotaHierarchyInfo[quotaInfo.Name] = make(map[string]struct{})
+	qt.quotaHierarchyInfo[quotaInfo.ParentName][quotaInfo.Name] = struct{}{}
+	return nil
+}
+
+func (qt *quotaTopology) UpdateQuota(oldQuota, newQuota *v1alpha1.ElasticQuota) error {
+	if newQuota == nil {
+		return fmt.Errorf("AddQuota param is nil")
+	}
+
+	if oldQuota != nil && reflect.DeepEqual(quotaFieldsCopy(oldQuota), quotaFieldsCopy(newQuota)) {
+		return nil
+	}
+
+	quotaName := newQuota.Name
+	if quotaName == extension.SystemQuotaName || quotaName == extension.RootQuotaName {
+		return fmt.Errorf("can not update quotaGroup :%v", quotaName)
+	}
+
+	qt.lock.Lock()
+	defer qt.lock.Unlock()
+
+	oldQuotaInfo, exist := qt.quotaInfoMap[quotaName]
+	if !exist {
+		return fmt.Errorf("quota not exist in quotaInfoMap:%v", quotaName)
+	}
+
+	if err := qt.validateQuotaSelfItem(newQuota); err != nil {
+		return err
+	}
+
+	newQuotaInfo := core.NewQuotaInfoFromQuota(newQuota)
+	if err := qt.validateQuotaTopology(newQuotaInfo); err != nil {
+		return err
+	}
+
+	// interchange between parentQuotaGroup and childQuotaGroup is prohibited now.
+	if oldQuotaInfo.IsParent != newQuotaInfo.IsParent {
+		return fmt.Errorf("IsParent is forbidden modify now, quotaName:%v", oldQuotaInfo.Name)
+	}
+
+	qt.quotaInfoMap[quotaName] = newQuotaInfo
+	if oldQuotaInfo.ParentName != newQuotaInfo.ParentName {
+		delete(qt.quotaHierarchyInfo[oldQuotaInfo.ParentName], oldQuotaInfo.Name)
+		qt.quotaHierarchyInfo[newQuotaInfo.ParentName][newQuotaInfo.Name] = struct{}{}
```

Comment:
this is a terrible bug,  1->2->3,  when we move 2's parent:   11->2->3,  you will lost 3 in qt.quotaHierarchyInfo[2]